src/ripple/app/consensus/RCLValidations.cpp

```diff
@@ -0,0 +1,337 @@
+
+    // Ensure validation is marked as trusted if signer currently trusted
+    auto pubKey = app_.validators().getTrustedKey(signer);
+    if (!val->isTrusted() && pubKey)
+        val->setTrusted();
+
+    if (!val->isTrusted())
+    {
+        JLOG(j_.trace()) << "Node "
+                         << toBase58(TokenType::TOKEN_NODE_PUBLIC, signer)
+                         << " not in UNL st="
+                         << val->getSignTime().time_since_epoch().count()
+                         << ", hash=" << hash
+                         << ", shash=" << val->getSigningHash()
+                         << " src=" << source;
+    }
+
+    // If not currently trusted, see if signer is currently listed
+    if (!pubKey)
+        pubKey = app_.validators().getListedKey(signer);
+
+    bool currVal = isCurrent(val);
+    // only add trusted or listed
+    if (currVal && (val->isTrusted() || pubKey))
+    {
+        ScopedLockType sl(lock_);
+        auto res = Validations::add(now(), *pubKey, val);
+
+        // This is a duplicate validadtion
+        if (res == AddOutcome::repeat)
+            return false;
+        // This validation replaced a prior one with the same sequence number
+        if (res == AddOutcome::sameSeq)
+        {
+            auto const seq = val->getFieldU32(sfLedgerSequence);
+            JLOG(j_.warn())
+                << "Trusted node "
+                << toBase58(TokenType::TOKEN_NODE_PUBLIC, *pubKey)
+                << " published multiple validations for ledger " << seq;
+        }
+
+        currVal = res == AddOutcome::current;
+    }
+
+    JLOG(j_.debug()) << "Val for " << hash << " from "
+                     << toBase58(TokenType::TOKEN_NODE_PUBLIC, signer)
+                     << " added "
+                     << (val->isTrusted() ? "trusted/" : "UNtrusted/")
+                     << (currVal ? "current" : "stale");
+
+    if (val->isTrusted() && currVal)
+    {
+        app_.getLedgerMaster().checkAccept(
+            hash, val->getFieldU32(sfLedgerSequence));
+        return true;
+    }
+
+    // FIXME: This never forwards untrusted validations
+    return false;
+}
+
+bool
+RCLValidations::isCurrent(STValidation::ref val)
+{
+    using ripple::isCurrent;
+    return isCurrent(parms(), now(), val->getSignTime(), val->getSeenTime());
+}
+
+std::vector<STValidation::pointer>
+RCLValidations::getTrustedForLedger(uint256 const& ledger)
+{
+    std::vector<STValidation::pointer> res;
+    {
+        ScopedLockType sl(lock_);
+        if (auto map = byLedger(ledger))
+        {
+            for (auto const& it : *map)
+                if (it.second.trusted())
+                    res.emplace_back(it.second.val_);
+        }
+    }
+    return res;
+}
+
+std::vector<NetClock::time_point>
+RCLValidations::getValidationTimes(uint256 const& ledger)
+{
+    std::vector<NetClock::time_point> times;
+    {
+        ScopedLockType sl(lock_);
+        if (auto map = byLedger(ledger))
```

Comment:
This is an important point and I agree. I tend to overuse auto when developing intially while the types are in flux. But readability for the future is essential. I will go back and fix.